geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/GeoMessage.scala

```diff
@@ -0,0 +1,184 @@
+
+  val id = feature.getID
+}
+
+/** Delete an existing [[SimpleFeature]]
+  *
+  * @param id the id of the simple feature
+ */
+case class Delete(override val timestamp: Instant, id: String) extends GeoMessage
+
+/** Delete all [[SimpleFeature]]s
+  */
+case class Clear(override val timestamp: Instant) extends GeoMessage
+
+
+/** Encodes [[GeoMessage]]s.  [[Clear]] and [[Delete]] messages are handled directly.  See class
+  * [[GeoMessageEncoder]] for encoding [[CreateOrUpdate]] messages.
+  *
+  * The following encoding is used:
+  *
+  * Key: version (1 byte) type (1 byte) timstamp (8 bytes)
+  *
+  * The current version is 1
+  * The type is 'C' for create or update, 'D' for delete and 'X' for clear
+  *
+  * Value
+  *   CreateOrUpdate: serialized simple feature
+  *   Delete: simple feature ID as UTF-8 bytes
+  *   Clear: empty
+  *
+  */
+object GeoMessageEncoder {
+
+  val version: Byte = 1
+  val createOrUpdateType: Byte = 'C'
+  val deleteType: Byte = 'D'
+  val clearType: Byte = 'X'
+
+  private val EMPTY = Array.empty[Byte]
+
+  def encodeKey(msg: GeoMessage): Array[Byte] = {
+
+    val msgType: Byte = msg match {
+      case c: CreateOrUpdate => createOrUpdateType
+      case d: Delete => deleteType
+      case x: Clear => clearType
+      case u => throw new IllegalArgumentException(s"Invalid message: '$u'")
+    }
+
+    val bb = ByteBuffer.allocate(10)
+    bb.put(version)
+    bb.put(msgType)
+    bb.putLong(msg.timestamp.getMillis)
+    bb.array()
+  }
+
+  def encodeClearMessage(msg: Clear): Array[Byte] = EMPTY
+
+  def encodeDeleteMessage(msg: Delete): Array[Byte] = msg.id.getBytes(StandardCharsets.UTF_8)
+}
+
+/** Encodes [[GeoMessage]]s.
+  *
+  * @param schema the [[SimpleFeatureType]]; required to serialize [[CreateOrUpdate]] messages
+  */
+class GeoMessageEncoder(schema: SimpleFeatureType) {
+
+  import GeoMessageEncoder._
+
+  val serializer: SimpleFeatureSerializer = new KryoFeatureSerializer(schema, SerializationOptions.withUserData)
+
+  def encodeMessage(msg: GeoMessage): Array[Byte] = msg match {
+    case c: CreateOrUpdate =>
+      encodeCreateOrUpdateMessage(c)
+    case d: Delete =>
+      encodeDeleteMessage(d)
+    case x: Clear =>
+      encodeClearMessage(x)
+    case u =>
+      throw new IllegalArgumentException(s"Invalid message: '$u'")
+  }
+
+  def encodeCreateOrUpdateMessage(msg: CreateOrUpdate): Array[Byte] = serializer.serialize(msg.feature)
+}
+
+/** Decodes an encoded [[GeoMessage]].
+  *
+  * @param schema the [[SimpleFeatureType]]; required to deserialize [[CreateOrUpdate]] messages
+  */
+class GeoMessageDecoder(schema: SimpleFeatureType) extends Logging {
+
+  case class MsgKey(version: Byte, msgType: Byte, ts: Instant)
+
+  val serializer: SimpleFeatureDeserializer = new KryoFeatureSerializer(schema, SerializationOptions.withUserData)
+
+  def decode(key: Array[Byte], msg: Array[Byte]): GeoMessage = {
+    val MsgKey(version, msgType, ts) = decodeKey(key)
```

Comment:
Is MsgKey only used to encode/extract for this line?  If so, a Tuple3 should suffice.